packages/custom-functions-metadata/src/parseTree.ts

```diff
@@ -0,0 +1,1114 @@
+    errorString = `The custom function name is too long. It must be 128 characters or less.`;
+    extra.errors.push(logError(errorString, position));
+  }
+}
+
+/**
+ * Normalize the id of the custom function
+ * @param id Parameter id of the custom function
+ */
+function normalizeCustomFunctionId(id: string): string {
+  return id ? id.toLocaleUpperCase() : id;
+}
+
+/**
+ * Determines the options parameters for the json
+ * @param func - Function
+ * @param isStreamingFunction - Is is a steaming function
+ */
+function getOptions(
+  func: ts.FunctionDeclaration,
+  isStreamingFunction: boolean,
+  isCancelableFunction: boolean,
+  isInvocationFunction: boolean,
+  extra: IFunctionExtras
+): IFunctionOptions {
+  const optionsItem: IFunctionOptions = {
+    cancelable: isCancelableTag(func, isCancelableFunction),
+    requiresAddress: isAddressRequired(func),
+    stream: isStreaming(func, isStreamingFunction),
+    volatile: isVolatile(func),
+    requiresParameterAddresses: isRequiresParameterAddresses(func),
+  };
+
+  if (optionsItem.requiresAddress || optionsItem.requiresParameterAddresses) {
+    let errorParam: string = optionsItem.requiresAddress ? "@requiresAddress" : "@requiresParameterAddresses";
+
+    if (!isStreamingFunction && !isCancelableFunction && !isInvocationFunction) {
+      const functionPosition = getPosition(func, func.parameters.end);
+      const errorString = `Since ${errorParam} is present, the last function parameter should be of type CustomFunctions.Invocation :`;
+      extra.errors.push(logError(errorString, functionPosition));
+    }
+
+    if (isStreamingFunction) {
+      const functionPosition = getPosition(func);
+      const errorString = `${errorParam} cannot be used with @streaming.`;
+      extra.errors.push(logError(errorString, functionPosition));
+    }
+  }
+
+  return optionsItem;
+}
+
+/**
+ * Determines the results parameter for the json
+ * @param func - Function
+ * @param isStreaming - Is a streaming function
+ * @param lastParameter - Last parameter of the function signature
+ */
+function getResults(
+  func: ts.FunctionDeclaration,
+  isStreamingFunction: boolean,
+  lastParameter: ts.ParameterDeclaration,
+  jsDocParamTypeInfo: { [key: string]: IJsDocParamType },
+  extra: IFunctionExtras,
+  enumList: string[]
+): IFunctionResult {
+  let resultType = "any";
+  let resultDim = "scalar";
+  const defaultResultItem: IFunctionResult = {
+    dimensionality: resultDim,
+    type: resultType,
+  };
+
+  const lastParameterPosition = getPosition(lastParameter);
+
+  // Try and determine the return type.  If one can't be determined we will set to any type
+  if (isStreamingFunction) {
+    const lastParameterType = lastParameter.type as ts.TypeReferenceNode;
+    if (!lastParameterType) {
+      // Need to get result type from param {type}
+      const name = (lastParameter.name as ts.Identifier).text;
+      const ptype = jsDocParamTypeInfo[name];
+      // @ts-ignore
```

Comment:
Copied from original: `@ts-ignore` is generally something to avoid. Generally, this should be flagged by linter rules.